server/keyspace/keyspace_test.go

```diff
@@ -0,0 +1,229 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package keyspace
+
+import (
+	"fmt"
+	"math/rand"
+	"strconv"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/pingcap/kvproto/pkg/keyspacepb"
+	"github.com/stretchr/testify/require"
+	"github.com/tikv/pd/pkg/mock/mockid"
+	"github.com/tikv/pd/server/storage/endpoint"
+	"github.com/tikv/pd/server/storage/kv"
+)
+
+const testConfig = "test config"
+
+func newKeyspaceManager() *Manager {
+	store := endpoint.NewStorageEndpoint(kv.NewMemoryKV(), nil)
+	allocator := mockid.NewIDAllocator()
+	return NewKeyspaceManager(store, allocator)
+}
+
+func createKeyspaceRequests(count int) []*CreateKeyspaceRequest {
+	now := time.Now()
+	requests := make([]*CreateKeyspaceRequest, count)
+	for i := 0; i < count; i++ {
+		requests[i] = &CreateKeyspaceRequest{
+			Name: fmt.Sprintf("test_keyspace%d", i),
+			InitialConfig: map[string]string{
+				"config_entry_1": "100",
+				"config_entry_2": "200",
+			},
+			Now: now,
+		}
+	}
+	return requests
+}
+func TestCreateKeyspace(t *testing.T) {
+	re := require.New(t)
+	manager := newKeyspaceManager()
+	requests := createKeyspaceRequests(10)
+
+	for i, request := range requests {
+		created, err := manager.CreateKeyspace(request)
+		re.NoError(err)
+		re.Equal(uint32(i+1), created.Id)
+		matchCreateRequest(re, request, created)
+
+		loaded, err := manager.LoadKeyspace(request.Name)
+		re.NoError(err)
+		re.Equal(uint32(i+1), loaded.Id)
+		matchCreateRequest(re, request, loaded)
+	}
+
+	// create a keyspace with existing name must return error
+	_, err := manager.CreateKeyspace(requests[0])
+	re.Error(err)
+
+	// create a keyspace with empty name must return error
+	_, err = manager.CreateKeyspace(&CreateKeyspaceRequest{Name: ""})
+	re.Error(err)
+}
+
+func TestUpdateKeyspace(t *testing.T) {
+	re := require.New(t)
+	manager := newKeyspaceManager()
+	requests := createKeyspaceRequests(5)
+	for i, createRequest := range requests {
+		_, err := manager.CreateKeyspace(createRequest)
+		re.NoError(err)
+		updateRequest := &UpdateKeyspaceRequest{
+			Name:        createRequest.Name,
+			UpdateState: true,
+			NewState:    keyspacepb.KeyspaceState(rand.Int31n(3)),
+			Now:         time.Now(),
+			ToPut: map[string]string{
+				"config_entry_1":    strconv.Itoa(i),
+				"config_entry_2":    strconv.Itoa(i),
+				"additional_config": strconv.Itoa(i),
+			},
+			ToDelete: []string{"config_entry_2"},
+		}
+
+		updated, err := manager.UpdateKeyspace(updateRequest)
+		re.NoError(err)
+		matchUpdateRequest(re, updateRequest, updated)
+	}
+}
+
+func TestLoadRangeKeyspace(t *testing.T) {
+	re := require.New(t)
+	manager := newKeyspaceManager()
+	requests := createKeyspaceRequests(10)
+
+	startID := 100
+	// force keyspace id start with startID
+	for i := 0; i < startID-1; i++ {
+		_, _ = manager.idAllocator.Alloc()
```

Comment:
How about add a function to set startid when init a new allocator? Or it can be add as a TODO task.